### PR TITLE
Add more array.sort test combination

### LINKED ISSUE #41376: Add more Array.Sort test combinations, per #41234
Should we have more test combinations, given this involves the complexity of pivots, etc?

_Originally posted by @danmosemsft in https://github.com/dotnet/runtime/pull/41234#issuecomment-678879413_

### PR DESCRIPTION
**Pull request by DDolG**
Fixes: #41376

*Added tests for sorting arrays

*Separated tests for integer from string

*Added comments for integers and strings

EDIT (@krwq): Added `Fixes ...`

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=68489) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=68489) before we can accept your contribution.<br/><br/>:x: DDolG [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=68489)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=68489) it.</sub>

**Comment by krwq** [comment]
@DDolG I assume this is fixing https://github.com/dotnet/runtime/issues/41376 ? If so can you modify your PR description so that first line says "Fixes #41376" (this way merging PR automatically closes the issue)

**Comment by DDolG** [comment]
> @DDolG I assume this is fixing #41376 ? If so can you modify your PR description so that first line says "Fixes #41376" (this way merging PR automatically closes the issue)

I assume this is fixing #41376 ? Yes.
Thank you very much.
Thank you very much for your last comment. Very concise and understandable. I am very grateful to You.

File: src/libraries/System.Runtime/tests/System/ArrayTests.cs [hunk 1]
```diff
@@ -3331,34 +3331,44 @@ public static IEnumerable<object[]> Sort_Array_TestData()
 
         public static IEnumerable<object[]> Sort_SZArray_TestData()
         {
+			// Int
```
Comment (krwq) [inline src/libraries/System.Runtime/tests/System/ArrayTests.cs:3334]: nit:
```suggestion
			    // Int
```

File: src/libraries/System.Runtime/tests/System/ArrayTests.cs [hunk 2]
```diff
@@ -3331,34 +3331,44 @@ public static IEnumerable<object[]> Sort_Array_TestData()
...
-
             yield return new object[] { new int[] { 5, 2, 9, 8, 4, 3, 2, 4, 6 }, 0, 0, null, new int[] { 5, 2, 9, 8, 4, 3, 2, 4, 6 } };
             yield return new object[] { new int[] { 5, 2, 9, 8, 4, 3, 2, 4, 6 }, 9, 0, null, new int[] { 5, 2, 9, 8, 4, 3, 2, 4, 6 } };
+            yield return new object[] { new int[] { 1, 2, 3, 4 }, 0, 4, null, new int[] { 1, 2, 3, 4 } };
+            yield return new object[] { new int[] { 4, 3, 2, 1 }, 0, 4, null, new int[] { 1, 2, 3, 4 } };
+            yield return new object[] { new int[] { 4, 3, 2, 1 }, 1, 2, null, new int[] { 4, 2, 3, 1 } };
+            yield return new object[] { new int[] { 4, 3, 2 }, 0, 3, new IntegerComparer(), new int[] { 2, 3, 4 } };
+            yield return new object[] { new int[] { 4, 3, 2 }, 0, 3, null, new int[] { 2, 3, 4 } };
 
+			// String
```
Comment (krwq) [inline src/libraries/System.Runtime/tests/System/ArrayTests.cs:3351]:
```suggestion
			     // String
```

**Comment by DDolG** [comment]
> @DDolG I assume this is fixing #41376 ? If so can you modify your PR description so that first line says "Fixes #41376" (this way merging PR automatically closes the issue)

Krzysztof I am very grateful to you for the feedback on the code.
Krzysztof, please tell me what these 6 spaces are for. I read the [coding guide](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md), it doesn't say anything about it.

**Comment by krwq** [comment]
this was to align indentations of comments and code so they are the same across the file (in the very end of the coding style doc there is `When editing files, keep new code and changes consistent with the style in the files.` - not sure why it's in the very end considering it's probably one of the most important style rule 😄)

**Comment by DDolG** [comment]
> Thanks a lot. I began to understand.

**Comment by krwq** [comment]
I was thinking of good way to check this PR and here are my conclusions:
- this PR does improve coverage
- there is still lots of room for improvement here: i.e. we never hit heap sort cases for large arrays

Here is how you can check what's covered and what's not locally (after running `build.cmd -rc Release -s clr+libs` at least once from root of the repo):
- go to `src\libraries\System.Runtime\tests`
- run `dotnet test /p:Outerloop=true /p:Coverage=true`
- after it's finished running somewhere in the bottom of the output you should see something along the lines of:

```
  2022-04-26T16:24:27: Writing report file 'd:\src\runtime\artifacts\bin\System.Runtime.Tests\Debug\net7.0-windows\TestResults\report\index.html'
  2022-04-26T16:24:28: Report generation took 8.5 seconds
```

- open the path from the log, in my case it's `d:\src\runtime\artifacts\bin\System.Runtime.Tests\Debug\net7.0-windows\TestResults\report\index.html`
- search for `System.Array` and click on link or alternatively directly open `System.Private.CoreLib_Array.html` from the same folder
- search for method you'd like to increase coverage for (scroll to the bottom/search for or click on the method name on the right)
- after you're done editing the test re-run the `dotnet test ....` command to re-generate report (might be a good idea to copy original report first for comparison) - make sure you've actually increased the coverage you planned

If you lack time I'm ok with merging this as is. Please let me know if you plan to further update the PR.

**Comment by DDolG** [comment]
error create log 2
Unable to start testing with log generation.
I think to allocate sorting in a separate program to add logging there. And then form such a sequence that the HeapSort method is involved. Unfortunately, nothing better comes to mind.
Please tell me, if the sequence is large enough, will it need to be placed in another place?

**Comment by krwq** [comment]
@DDolG you;re seeing two issues there:
- test failure - most likely caused by timeout (I cannot see the full log to tell which test) - it's possible this is preventing coverage report from being generated - I think this problem might be related to your local changes - try fixing it (or temporarily disabling failing test) and see if report gets generated before you try anything else
- `Html Logger Error` - I'm not sure what that is but I think I've sometimes seen that locally and I believe it can be ignored

Can you save your local changes try doing `git clean -fdx` from the root of the repo, rebuild everything (`build -rc Release -s clr+libs`) and run test again with `/p:Coverage=true` and see if that still happens?

> Please tell me, if the sequence is large enough, will it need to be placed in another place?

I'm not sure what sequence are you referring to. In general debugging through test might help you figure out how to hit specific test (it might not be super trivial because of all of the if statements in the sorting code but it likely will be easier than guessing) - simply open project in VS and on the test explorer window search for test and right click and click Debug (make sure to put a breakpoint in the test first). Also if test is `[Theory]` you might want to comment out all cases but one since they run in parallel and that might make debugging much harder otherwise

**Comment by DDolG** [comment]
Thank a lot. Awesome way to check test coverage.
Krzysztof я увидел что есть методы, которые не покрыты тестами.
I will try to cover these methods with tests.
Thank you very much for teaching me this technology.
image

**Comment by DDolG** [comment]
Hello. Help me please. I have the impression that I'm not testing correctly.
I do testing according to the method written by Krzysztof. The output is a report. This report shows that the ArraySortHelper class is not covered by tests. Although if you separately debug each test, you can see that it enters the methods of this class.
image
[System.Private.CoreLib_ArraySortHelper_1.zip](https://github.com/dotnet/runtime/files/8590289/System.Private.CoreLib_ArraySortHelper_1.zip)

**Review by jeffhandley — APPROVED** [review]
> Hello. Help me please. I have the impression that I'm not testing correctly.

Sorry we've not been able to jump back in and help here, @DDolG. Per @krwq's earlier comment, it's OK to merge this as-is without adding further coverage at this time though.

**Comment by jeffhandley** [comment]
Thanks for the contribution, @DDolG!

**Comment by DDolG** [comment]
@jeffhandley thanks a lot. I am very grateful to you for the trust you have placed in me.

**Pull request merged**